<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Ejercicio-Fechas-Publicación" data-toc-modified-id="Ejercicio-Fechas-Publicación-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Ejercicio Fechas Publicación</a></span></li><li><span><a href="#Ejercicio-Tipo-de-Contrato-y-Subtipo-de-Contrato" data-toc-modified-id="Ejercicio-Tipo-de-Contrato-y-Subtipo-de-Contrato-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Ejercicio Tipo de Contrato y Subtipo de Contrato</a></span></li><li><span><a href="#Ejercicio-Duración-del-contrato" data-toc-modified-id="Ejercicio-Duración-del-contrato-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ejercicio Duración del contrato</a></span></li></ul></div>

# Ejercicios Limpieza de Datos <a class="tocSkip">

Lady Tatiana Arévalo López  Cod:2213306

Taller 1 segundo corte

29 de marzo de 2022

visualización para ciencia de datos en Python

In [ ]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None) #Mostrar todas las columnas

## Importar datos
Datos extraídos de la fuente 'SECOP II - Procesos de Contratación'

Fecha creación: 30 de septiembre de 2019

Datos actualizados por última vez: 31 de enero de 2021

Fuente: Datos SECOP

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
df_SECOP = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/SECOP_II_-_Procesos_de_Contrataci_n.csv', 
                sep = ',',
                 thousands=',', decimal='.',
                 skip_blank_lines=True,
                 low_memory = False,
                 dtype={
                     'Nit Entidad': str
                     })
df_SECOP.head()

Entidad Nit Entidad  \
0                     DISPENSARIO MEDICO DE MEDELLÍN  8300396705   
1  DIRECCION DE SANIDAD EJERCITO DISPENSARIO MEDI...   830039670   
2                                  INDUSTRIA MILITAR   899999044   
3        MINISTERIO DE VIVIENDA. CIUDAD Y TERRITORIO   900463725   
4                                   HOSPITAL CENTRAL   830067597   

         Departamento Entidad Ciudad Entidad OrdenEntidad  \
0                   Antioquia       Medellín     Nacional   
1                    Casanare          Yopal     Nacional   
2  Distrito Capital de Bogotá         Bogotá     Nacional   
3  Distrito Capital de Bogotá         Bogotá     Nacional   
4  Distrito Capital de Bogotá         Bogotá     Nacional   

  Entidad Centralizada   ID del Proceso       Referencia del Proceso  \
0                   Si   CO1.REQ.375424          021-DGSM-DMMED-2018   
1                   No   CO1.REQ.319641  078-CENACYOPAL-SANIDAD-2018   
2                   Si  CO1.REQ.1608416                     584/2020   
3                   No   CO1.REQ.306317             MVCT-CD-047-2018   
4                   No   CO1.REQ.728151        PN HOCEN CD 1178 2018   

            PCI ID del Portafolio  \
0  15-01-11-013   CO1.BDOS.366226   
1  15-01-11-024   CO1.BDOS.311763   
2            ND  CO1.BDOS.1559487   
3      40-01-01   CO1.BDOS.298411   
4  16-01-02-001   CO1.BDOS.648821   

                            Nombre del Procedimiento  \
0  PRESTACIÓN DE SERVICIOS MÉDICOS HOSPITALARIOS ...   
1                                         ODONTOLOGO   
2  Prestación del Servicio de Evaluaciones Médica...   
3                      CPSAG LUIS CARLOS ORIZ NOREÑA   
4                              PN HOCEN CD 1178 2018   

                       Descripción del Procedimiento                    Fase  \
0  PRESTACIÓN DE SERVICIOS MÉDICOS HOSPITALARIOS ...  Presentación de oferta   
1  PRESTACIÓN DE SERVICIOS PROFESIONALES COMO ODO...  Presentación de oferta   
2  Prestación del Servicio de Evaluaciones Médica...  Presentación de oferta   
3  Apoyar al Ministerio de Vivienda, Ciudad y Ter...  Presentación de oferta   
4  PRESTACION DE SERVICIOS DE APOYO A LA GESTION ...  Presentación de oferta   

  Fecha de Publicacion del Proceso Fecha de Ultima Publicación  \
0                       03/02/2018                  03/02/2018   
1                       01/19/2018                  01/19/2018   
2                       11/13/2020                  11/13/2020   
3                       01/13/2018                  01/12/2018   
4                       01/31/2019                  01/31/2019   

  Fecha de Publicacion (Fase Planeacion Precalificacion)  \
0                                                NaN       
1                                                NaN       
2                                                NaN       
3                                                NaN       
4                                                NaN       

  Fecha de Publicacion (Fase Seleccion Precalificacion)  \
0                                                NaN      
1                                                NaN      
2                                                NaN      
3                                                NaN      
4                                                NaN      

  Fecha de Publicacion (Manifestacion de Interes)  \
0                                             NaN   
1                                             NaN   
2                                             NaN   
3                                             NaN   
4                                             NaN   

  Fecha de Publicacion (Fase Borrador) Fecha de Publicacion (Fase Seleccion)  \
0                                  NaN                            03/02/2018   
1                                  NaN                            01/19/2018   
2                                  NaN                            11/13/2020   
3                                  NaN     

Observamos las variables de la base de datos para identificar las que vamos a necesitar en el proceso del taller.

In [ ]:
df_SECOP.dtypes

Entidad                                                   object
Nit Entidad                                               object
Departamento Entidad                                      object
Ciudad Entidad                                            object
OrdenEntidad                                              object
Entidad Centralizada                                      object
ID del Proceso                                            object
Referencia del Proceso                                    object
PCI                                                       object
ID del Portafolio                                         object
Nombre del Procedimiento                                  object
Descripción del Procedimiento                             object
Fase                                                      object
Fecha de Publicacion del Proceso                          object
Fecha de Ultima Publicación                               object
Fecha de Publicacion (Fas

## 1. Ejercicio Fechas Publicación



Teniendo en cuenta que existe una Fecha de Publicación del Proceso y una fecha de última publicación, realice la limpieza de formato de dichas variables. Defina una nueva variable en la que se calcule la diferencia de éstas dos fechas y responda si incluiría las dos fechas dentro de la fuente de datos.

*Utilice estadísticas descriptivas para responder la pregunta*

## **Solución:** 

Procedemos a la limpieza del formato de la variable fecha de Publicación del Proceso y una fecha de última publicación de la base de datos Datos_SECOP:

### 1. Limpieza y ejecución


In [27]:
df_SECOP['Fecha de Publicacion del Proceso'] = pd.to_datetime(df_SECOP['Fecha de Publicacion del Proceso'],format='%m/%d/%Y')
df_SECOP['Fecha de Publicacion del Proceso'].head()

0   2018-03-02
1   2018-01-19
2   2020-11-13
3   2018-01-13
4   2019-01-31
Name: Fecha de Publicacion del Proceso, dtype: datetime64[ns]

In [28]:
df_SECOP['Fecha de Ultima Publicación'] = pd.to_datetime(df_SECOP['Fecha de Ultima Publicación'],format='%m/%d/%Y')
df_SECOP['Fecha de Ultima Publicación'].head()

0   2018-03-02
1   2018-01-19
2   2020-11-13
3   2018-01-12
4   2019-01-31
Name: Fecha de Ultima Publicación, dtype: datetime64[ns]

Procedemos a seleccionar las dos columnas con las variables Fecha de Publicacion del Proceso y Fecha de Ultima Publicación y creo una base que se llama datos_Fechas para observar el analisis para las fechas de publicación.

In [29]:
columns = ['Entidad', 'Fecha de Publicacion del Proceso','Fecha de Ultima Publicación']
datos_Fechas = df_SECOP.copy()
datos_Fechas = datos_Fechas[columns]
datos_Fechas

Entidad  \
0                          DISPENSARIO MEDICO DE MEDELLÍN   
1       DIRECCION DE SANIDAD EJERCITO DISPENSARIO MEDI...   
2                                       INDUSTRIA MILITAR   
3             MINISTERIO DE VIVIENDA. CIUDAD Y TERRITORIO   
4                                        HOSPITAL CENTRAL   
...                                                   ...   
876763                              MINISTERIO DE CULTURA   
876764                 SENA HOTELERIA TURISMO Y ALIMENTOS   
876765                            MINISTERIO DEL INTERIOR   
876766                               PERSONERIA DE BOGOTA   
876767  COMPLEJO CARCELARIO Y PENITENCIARIO DE IBAGUE ...   

       Fecha de Publicacion del Proceso Fecha de Ultima Publicación  
0                            2018-03-02                  2018-03-02  
1                            2018-01-19                  2018-01-19  
2                            2020-11-13                  2020-11-13  
3                            2018-01-13                  2018-01-12  
4                            2019-01-31                  2019-01-31  
...                                 ...                         ...  
876763                       2020-06-03                  2020-06-03  
876764                       2020-03-02                  2020-03-02  
876765                       2021-02-02                  2021-02-01  
876766                       2019-02-19                  2019-02-19  
876767                       2019-10-15                  2019-10-15  

[876768 rows x 3 columns]

In [30]:
datos_Fechas['dif_fechas'] =  (datos_Fechas['Fecha de Publicacion del Proceso'] - datos_Fechas['Fecha de Ultima Publicación']).dt.days 
datos_Fechas

Entidad  \
0                          DISPENSARIO MEDICO DE MEDELLÍN   
1       DIRECCION DE SANIDAD EJERCITO DISPENSARIO MEDI...   
2                                       INDUSTRIA MILITAR   
3             MINISTERIO DE VIVIENDA. CIUDAD Y TERRITORIO   
4                                        HOSPITAL CENTRAL   
...                                                   ...   
876763                              MINISTERIO DE CULTURA   
876764                 SENA HOTELERIA TURISMO Y ALIMENTOS   
876765                            MINISTERIO DEL INTERIOR   
876766                               PERSONERIA DE BOGOTA   
876767  COMPLEJO CARCELARIO Y PENITENCIARIO DE IBAGUE ...   

       Fecha de Publicacion del Proceso Fecha de Ultima Publicación  \
0                            2018-03-02                  2018-03-02   
1                            2018-01-19                  2018-01-19   
2                            2020-11-13                  2020-11-13   
3                            2018-01-13                  2018-01-12   
4                            2019-01-31                  2019-01-31   
...                                 ...                         ...   
876763                       2020-06-03                  2020-06-03   
876764                       2020-03-02                  2020-03-02   
876765                       2021-02-02                  2021-02-01   
876766                       2019-02-19                  2019-02-19   
876767                       2019-10-15                  2019-10-15   

        dif_fechas  
0              0.0  
1              0.0  
2              0.0  
3              1.0  
4              0.0  
...            ...  
876763         0.0  
876764         0.0  
876765         1.0  
876766         0.0  
876767         0.0  

[876768 rows x 4 columns]

Vamos a realizar un analisis descriptivo basico para evaluar nuestra nueva variable y miramos la consistencia 

In [31]:
datos_Fechas['dif_fechas'].value_counts(dropna=False)

0.0    750871
1.0    125895
NaN         2
Name: dif_fechas, dtype: int64

In [33]:
datos_Fechas['porcentaje'] = (datos_Fechas['dif_fechas'] / 
                  datos_Fechas['dif_fechas'].sum()) * 10
datos_Fechas

Entidad  \
0                          DISPENSARIO MEDICO DE MEDELLÍN   
1       DIRECCION DE SANIDAD EJERCITO DISPENSARIO MEDI...   
2                                       INDUSTRIA MILITAR   
3             MINISTERIO DE VIVIENDA. CIUDAD Y TERRITORIO   
4                                        HOSPITAL CENTRAL   
...                                                   ...   
876763                              MINISTERIO DE CULTURA   
876764                 SENA HOTELERIA TURISMO Y ALIMENTOS   
876765                            MINISTERIO DEL INTERIOR   
876766                               PERSONERIA DE BOGOTA   
876767  COMPLEJO CARCELARIO Y PENITENCIARIO DE IBAGUE ...   

       Fecha de Publicacion del Proceso Fecha de Ultima Publicación  \
0                            2018-03-02                  2018-03-02   
1                            2018-01-19                  2018-01-19   
2                            2020-11-13                  2020-11-13   
3                            2018-01-13                  2018-01-12   
4                            2019-01-31                  2019-01-31   
...                                 ...                         ...   
876763                       2020-06-03                  2020-06-03   
876764                       2020-03-02                  2020-03-02   
876765                       2021-02-02                  2021-02-01   
876766                       2019-02-19                  2019-02-19   
876767                       2019-10-15                  2019-10-15   

        dif_fechas  porcentaje  
0              0.0    0.000000  
1              0.0    0.000000  
2              0.0    0.000000  
3              1.0    0.000079  
4              0.0    0.000000  
...            ...         ...  
876763         0.0    0.000000  
876764         0.0    0.000000  
876765         1.0    0.000079  
876766         0.0    0.000000  
876767         0.0    0.000000  

[876768 rows x 5 columns]

- Se crea una nueva variable que se llama dif_fechas donde ponemos la diferencia para así evaluar tiempo que tardo desde la Fecha de Publicacion del Proceso hasta la Fecha de Ultima Publicación y observar si tiene alguna diferencia significativa para tomarla o no en nuestro análisis.

- En conclusión, podemos observar que la diferencia entre las dos variables Fecha de Publicacion del Proceso y Fecha de Ultima Publicación no tiene ninguna relevancia la diferencia es de 1 día para 125895 entidades y tenemos cero diferencias en 750871 entidades, la diferencia porcentual es del 0.000079 % para todos los casos con 1 día de diferencia.

- Si incluiría las dos variables ya que hay varios datos que tiene n1 día de diferencia y veo que para el lector es importante saber que no hay una diferencia significativa en las fechas de Fecha de Publicación del Proceso y Fecha de Ultima Publicación.

## Ejercicio Tipo de Contrato y Subtipo de Contrato
De acuerdo con las variables Contrato y Subcontrato, realice una tabla cruzada y responda ¿Vale la pena dejar ambas variables en la fuente de datos?

A continuación, vamos a crear una tabla cruzada con nuestras dos variables Contrato y Subcontrato para evaluar si son representativas o no para dejarlas en la fuente de datos.

In [34]:
pd.crosstab(df_SECOP['Subtipo de Contrato'],df_SECOP['Tipo de Contrato'])

Tipo de Contrato     27 - Otros servicios  Acuerdo Marco  \
Subtipo de Contrato                                        
No Especificado                     42777           2782   

Tipo de Contrato     Alquiler de edificios  Arrendamiento Muebles  Comisión  \
Subtipo de Contrato                                                           
No Especificado                       7554                    527        84   

Tipo de Contrato     Compraventa  Concesión  Consultoría  Decreto 092 de 2017  \
Subtipo de Contrato                                                             
No Especificado            42412        108         4164                28466   

Tipo de Contrato     Interventoría     ND  No Especificado   Obra  Seguros  \
Subtipo de Contrato                                                          
No Especificado               4725  28827                5  11790     2218   

Tipo de Contrato     Servicios de aprovisionamiento  Suministros  \
Subtipo de Contrato                                                
No Especificado                              653206        46985   

Tipo de Contrato     Ventas de muebles  
Subtipo de Contrato                     
No Especificado                    138

Evaluando nuestra tabla cruzada no dejaria las variables pues se evidencia que con el no especificado podriamos manejar mejor la base de datos, los datos no quedarian tan robustos con las dos variables evaluadas.

## Ejercicio Duración del contrato
Teniendo en cuenta que la duración de los contratos tiene métricas diferentes, encuentre una estrategia para crear una nueva variable en la cuál se encuentre parametrizada la variable de duración del contrato.

In [ ]:
df_SECOP['Unidad de Duracion'].value_counts()

Dias     390312
Meses    358815
ND        26337
Años       6298
Name: Unidad de Duracion, dtype: int64

In [ ]:
df_SECOP[['Unidad de Duracion','Duracion']].sample(20)

Unidad de Duracion  Duracion
65723               Meses         5
187533              Meses        11
803114              Meses         4
556552              Meses         6
347984              Meses         6
244723               Dias       345
38169                Dias       330
624114              Meses         4
371379              Meses         7
384426               Dias       145
25772                Dias       180
688710               Dias        60
547263               Dias        30
111272               Dias       300
293092               Dias       359
467719                 ND         0
421116              Meses         2
680976              Meses         3
313880               Dias       347
716156               Dias        30